redash/tasks/queries/maintenance.py

```diff
@@ -52,9 +52,11 @@ def _should_refresh_query(query):
 
 def _apply_default_parameters(query):
```

Comment:
One issue with what we're doing here is that it changes the query text -> changes the query hash, which in the end will result in `latest_query_data_id` not being updated for this query.

It's not a huge issue, but can lead to some issues (like not showing when it was last refreshed). 

One way to solve it is to generate the query_hash of the Query object taking into account the limit part. This happens here:

https://github.com/getredash/redash/blob/205915e6db802d61667c14f8978d0c02fcccfe46/redash/models/__init__.py#L868-L871

@rauchy wdyt?

Reply:
I'm sorry for chiming in late here, but why is a function called `apply_default_parameters` responsible for applying limits? Also there's [a small DRY issue](https://github.com/getredash/redash/pull/5088/files/471586eabac9a23110d4751172e04944441ee887#diff-f734d32676ec526e17e5ae92d41b5787R82) with the same problem.

The way I see it - it looks like query runners are the ones who _know_ if they support limiting, and they are also the ones who _apply_ the limit, so it feels like limiting should be encapsulated inside them. Current situation is "asking, not telling": `if query_runner.supports_limiting then query_runner.apply_limit(text)`. I'd prefer a [Tell, Don't Ask](https://thoughtbot.com/blog/tell-dont-ask).

I have a feeling this design is caused by continuous creep of suboptimal design in which query runners work with query text, not query objects. This results in solution sprawl and, as @arikfr pointed out - collision with other parts of the system, such as query hashes.

My bottom line - if query runners own this behaviour, make the changes needed to provide them the information they need and let them perform limiting without changing anything outside the query runners. I believe this will also solve the hashing problem.

Reply:
According to @rauchy suggestion, I just made several changes to the code( moving most of the logic to query_runner etc). But the query hash won't be solved because every time we refresh, we are changing the query text and results in different hash. I am going to think about how to solve this and let me know if you have any ideas. Thanks!

Reply:
That's a good point, @rauchy. But such a refactor is out of scope for this PR. Also, I'm not 100% convinced that handling parameters should be in the Query Runner code. If we go this path, then even bigger refactor needed to make sure there is minimal duplication between the different query runner.

Also on its own it won't solve the query hash problem: it is the right behavior to assign a different hash for the limited query - otherwise we might update cache for wrong queries. Which leads me to believe that updating the query hash based on the current `apply_limit` value of the query is the right thing to do (at least given current design).

Reply:
Just updated the code with fix for query hash. I tested locally and queries are refreshed succesfully, let me know if you have any questions!